### PR TITLE
Add a simple size test

### LINKED ISSUE #80165: Bringing down the size of Hello World under 2 MB
Hello World with Native AOT is currently at around 2.7 MB. I spent a couple hours to see what would it take to bring it down to the same size as Golang (2 MB). It might actually be achievable with our default trimming settings and feature switches with a couple days of BCL/compiler work.

The benefits would be threefold:
* Our internal efficiency (compiling non-xunit-based tests will get faster and require less disk space, which is not insignificant as long as we have thousands of tiny tests)
* We can delete reflection disabled mode
* Bragging rights

The https://github.com/dotnet/runtime/commit/6686dbcf4654715b238bd359727a552ef8b968ad commit has a couple hacks in it around areas of interest. With those hacks, the size of a hello world goes down to 2.16 MB. With `InvariantGlobalization` and `UseSystemResourceKeys` it goes further down to 1.53 MB. This is awfully close to what is achievable with reflection disabled mode. We actually regressed the reflection disabled mode because a Hello World now has a generic virtual method call around enums and we need the entirety of the type loader to support that. But I digress.

Should we choose to accept this challenge, we'll need to fix all of the issues in the commit to really reap the benefit. The benefit is elimination of the entire reflection stack that deals with type members (in fact my goal was to get `MemberPolicies` class out of the hello world).

Here's the problems:
- [x] Stack frame should not eagerly try to populate the MethodBase, only once it's requested.
- [x] Type components cache should not try to eagerly populate the members. We use this class to stash details about enums in the GenericCache property. Populating anything else is a waste.
- [x] Type.IsEnum and IsValueType. We expand these as intrinsics in codegen for RuntimeTypes. Hello world only needs them for RuntimeTypes. IL scanner might need to learn the RyuJITs trick. (This is really only a problem because IL scanner runs. #79594 might already help with this to some extent but I didn't check. Would be nice if it did.)
- [x] Searching for DefaultDllImportSearchPathsAttribute is avoidable if this codepath is only reached from p/invoke resolution.
- [ ] ~~Attribute.Equals and GetHashCode. This is a tough one. .NET Native didn't use a reflection-based implementation. We could restore that. Other alternative is to see if we can eliminate all custom attribute descendants from a hello world. The theory is that if we make generation of custom attributes conditional on the presence of the code to read them, we might be able to get rid of them all. But not sure.~~ Chose not to address this and live with a small increase in size due to FieldInfo reflection. That one is less impactful than Method/ConstructorInfo.
- [X] Resource manager looking for attributes. I don't have ideas right now but I also didn't look at it in detail.

Cc @dotnet/ilc-contrib

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Resolves #80165.

The size of Hello World is currently 1.8 MB. This is achieved by avoiding any non-field reflection in the codepaths that are part of a hello world. If reflection comes into picture, the size jumps by several 100 kB because suddently we need a lot of code to support that. This is a smoke test to detect those situations. We're getting proper size testing in the dotnet/performance repo with trend histories, etc., but the invariant for Hello World is easy to check for and nice to gate commits on.

I chose the HwIntrinsics tests as a "victim" to host this because I don't want to introduce a new test for this.

Cc @dotnet/ilc-contrib

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
Hmm, the test infra is producing a 3.6 MB executable for ~Hello World with StripSymbols on Linux.

On my machine with Ubuntu 18.04, this is slightly under 1.9 MB.

A difference I see is that the test infra environment chose llvm-objcopy to do stripping whereas on my Ubuntu it chose objcopy.

@am11 did you see similar differences when you were working on this?

I guess we can just give the test more budget on Linux.

**Comment by am11** [comment]
Just like RyuJIT, the older versions of compiler toolchains (both llvm and gcc) are less efficient; llvm lesser efficient than gcc).

The only reason we have CentOS 7 around with old glibc, libstdc++ and llvm 9 is to enhance the official (portable) linux-x64 binary compatibility (glibc is backward-compatible). For size tests, I think we should use a modern platform https://github.com/dotnet/versions/blob/841708954543e7dffb8250e0561d06f753fdb97c/build-info/docker/image-info.dotnet-dotnet-buildtools-prereqs-docker-production.json#L2412 onward (ubuntu 22.04 with clang 14 or 15 in cross containers).

> A difference I see is that the test infra environment chose llvm-objcopy to do stripping whereas on my Ubuntu it chose objcopy.

Yup, llvm-objcopy wasn't that smart in v9, but in later versions, it is. If changing containers for smoke tests is not an option, then `<ObjCopyName>objcopy` will do the trick (override prop to force binutils' objcopy).

**Comment by ivanpovazan** [comment]
FWIW: we have noticed something similar in difference between XCode toolchain versions, more particularly `Xcode 14.1.0` vs `Xcode 13.3.0`, when tracking iOS application size with MonoAOT reported here: https://github.com/dotnet/runtime/issues/79879

File: src/tests/nativeaot/SmokeTests/HardwareIntrinsics/X64Baseline.csproj [hunk 1]
```diff
@@ -6,6 +6,7 @@
     <CLRTestTargetUnsupported Condition="'$(TargetArchitecture)' != 'x64'">true</CLRTestTargetUnsupported>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <DefineConstants>$(DefineConstants);BASELINE_INTRINSICS</DefineConstants>
+    <StripSymbols>true</StripSymbols>
```
Comment (MichalStrehovsky) [inline src/tests/nativeaot/SmokeTests/HardwareIntrinsics/X64Baseline.csproj:9]:
```suggestion
    <StripSymbols>true</StripSymbols>
    
    <!-- We should be able to delete once CI machines no longer use ancient LLVM -->
    <ObjCopyName>objcopy</ObjCopyName>
```

File: src/tests/nativeaot/SmokeTests/HardwareIntrinsics/x64NonVex.csproj [hunk 2]
```diff
@@ -6,6 +6,7 @@
     <CLRTestTargetUnsupported Condition="'$(TargetArchitecture)' != 'x64'">true</CLRTestTargetUnsupported>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <DefineConstants>$(DefineConstants);NON_VEX_INTRINSICS</DefineConstants>
+    <StripSymbols>true</StripSymbols>
```
Comment (MichalStrehovsky) [inline src/tests/nativeaot/SmokeTests/HardwareIntrinsics/x64NonVex.csproj:9]:
```suggestion
    <StripSymbols>true</StripSymbols>
    
    <!-- We should be able to delete once CI machines no longer use ancient LLVM -->
    <ObjCopyName>objcopy</ObjCopyName>
```

File: src/tests/nativeaot/SmokeTests/HardwareIntrinsics/x64Vex.csproj [hunk 3]
```diff
@@ -6,6 +6,7 @@
     <CLRTestTargetUnsupported Condition="'$(TargetArchitecture)' != 'x64'">true</CLRTestTargetUnsupported>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <DefineConstants>$(DefineConstants);VEX_INTRINSICS</DefineConstants>
+    <StripSymbols>true</StripSymbols>
```
Comment (MichalStrehovsky) [inline src/tests/nativeaot/SmokeTests/HardwareIntrinsics/x64Vex.csproj:9]:
```suggestion
    <StripSymbols>true</StripSymbols>
    
    <!-- We should be able to delete once CI machines no longer use ancient LLVM -->
    <ObjCopyName>objcopy</ObjCopyName>
```

**Comment by MichalStrehovsky** [comment]
> Yup, llvm-objcopy wasn't that smart in v9, but in later versions, it is. If changing containers for smoke tests is not an option, then `<ObjCopyName>objcopy` will do the trick (override prop to force binutils' objcopy).

Thanks, giving it a shot!

**Comment by MichalStrehovsky** [comment]
Looks like it did the trick. Digging in the CI logs, I found:

```
****************************************************
* Size test                                        *
* Size of the executable is   1,928 kB             *
****************************************************
```

File: src/tests/nativeaot/SmokeTests/HardwareIntrinsics/Program.cs [hunk 4]
```diff
@@ -12,6 +13,36 @@ static int Main()
...
+        Console.WriteLine($"* Size of the executable is {fileSize / 1024,7:n0} kB             *");
+        Console.WriteLine("****************************************************");
+
+        const int Meg = 1024 * 1024;
+        const int HalfMeg = Meg / 2;
+        long lowerBound, upperBound;
+        if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+        {
+            lowerBound = 2 * Meg; // 2 MB
+            upperBound = 4 * Meg; // 4 MB
```
Comment (am11) [inline src/tests/nativeaot/SmokeTests/HardwareIntrinsics/Program.cs:29]: Linux window is 1.5-2.0 MB. Should we shrink this to 3.0-3.5 MB?

Comment (MichalStrehovsky) [reply src/tests/nativeaot/SmokeTests/HardwareIntrinsics/Program.cs:29]: The OSX test will likely cross the 3.0 threshold downward soon. No point dealing with that - this is approximated by the Linux size test. OSX size is not in our 8.0 release criteria.

**Review by tlakollo — APPROVED** [review]
Looks good to me! :D

**Pull request merged**